packages/@aws-cdk/aws-cognito/README.md

```diff
@@ -546,6 +547,61 @@ pool.addClient('app-client', {
 });
 ```
 
+### Resource Servers
+
+A resource server is a server for access-protected resources. It handles authenticated requests from an app that has an
+access token. See [Defining Resource
+Servers](https://docs.aws.amazon.com/cognito/latest/developerguide/cognito-user-pools-define-resource-servers.html)
+for more information.
+
+An application may choose to model custom permissions via OAuth. Resource Servers provide this capability via custom scopes
+that are attached to an app client. The following example sets up a resource server for the 'users' resource for two different
+app clients and configures the clients to use these scopes.
+
+```ts
+const pool = new cognito.UserPool(this, 'Pool');
+
+const resourceServer = pool.addResourceServer('ResourceServer', {
+  identifier: 'users',
+  scopes: [
+    {
+      scopeName: 'read',
+      scopeDescription: 'Read-only access',
+    },
+    {
+      scopeName: '*',
+      scopeDescription: 'Full access',
+    },
+  ]
+});
+
+const readOnlyClient = pool.addClient('read-only-client', {
+  // ...
+  generateSecret: true,
+  oAuth: {
+    flows: {
+      clientCredentials: true,
+    },
+    scopes: [OAuthScope.custom('users/read')],
+  },
+});
+
+const fullAccessClient = pool.addClient('full-access-client', {
+  // ...
+  generateSecret: true,
+  oAuth: {
+    flows: {
+      clientCredentials: true,
+    },
+    scopes: [OAuthScope.custom('users/*')],
+  },
+});
+
+readOnlyClient.node.addDependency(resourceServer);
+fullAccessClient.node.addDependency(resourceServer);
```

Comment:
Wondering if we can do something better here.

Something like `OAuthScope.resourceServer(resourceServer, '*')`

This will make the integration between client and resource server more clear and obvious, and, will fix the dependency problem.

What do you think?

Reply:
So it's only for custom scopes that a resource server is needed. So ideally it'd be `Oauth.custom(resourceServer, '*')`.

But not sure how we'd not have duplicates from the scope level, cause if we pass in two custom scopes, then it'd do this dependency twice on the client.

So maybe something like this is a lot simpler:

```ts
userPool.addClient(scope, 'nah', {
  resourceServers: [resourceServer],
})
```

Reply:
`OAuthScope.custom(resourceServer, '*')` - this is a breaking change to the API so that's not going to work.

> But not sure how we'd not have duplicates from the scope level, cause if we pass in two custom scopes, then it'd do this dependency twice on the client.

Not entirely sure I understand what you mean. 
The dependency will be automatically derived by CloudFormation and the CDK since you will need to call `resourceServer.userPoolResourceServerId.`. You will not need to do anything extra, including calling `addDependency()`.

Reply:
> So maybe something like this is a lot simpler:
>
> ``` 
> userPool.addClient(scope, 'nah', {
>  resourceServers: [resourceServer],
> })
> ```

From what I understand, a resource server can have multiple scopes. A client can pick one or more of these. 
How would that work in this case?

Reply:
My choice of variable names might've been confusing. Here's a full example.

```ts
const pool = new cognito.UserPool(this, 'Pool');

const resourceServer = pool.addResourceServer('ResourceServer', {
  identifier: 'users',
  scopes: [
    {
      scopeName: 'read',
      scopeDescription: 'Read-only access',
    },
    {
      scopeName: '*',
      scopeDescription: 'Full access',
    },
  ]
});

const readOnlyClient = pool.addClient('read-only-client', {
  // ...
  generateSecret: true,
  oAuth: {
    flows: {
      clientCredentials: true,
    },
    resourceServers: [resourceServer],
    scopes: [OAuthScope.custom('users/read')],
  },
});

const fullAccessClient = pool.addClient('full-access-client', {
  // ...
  generateSecret: true,
  oAuth: {
    flows: {
      clientCredentials: true,
    },
    scopes: [OAuthScope.custom('users/*')],
    resourceServers: [resourceServer],
  },
});
```

Reply:
Any thought here @nija-at? Lemme know if to go ahead with the above

Reply:
I still think my original suggestion on this thread is better. 

It makes it clear that which scope is being used from which resource server. It also makes it clear that the resource server is being used to configure scope.

Reply:
You're referring to:

> OAuthScope.resourceServer(resourceServer, '*')

So the usage would be like the below?

```js
const fullAccessClient = pool.addClient('full-access-client', {
  // ...
  generateSecret: true,
  oAuth: {
    flows: {
      clientCredentials: true,
    },
    scopes: [
      OAuthScope.resourceServer(resourceServer, 'users/*'), 
      OAuthScope.resourceServer(resourceServer, 'users/read')
    ],
  },
});
```

So for implementation. In the UserPoolClient class. 
We have to check scopes, see which are of type `resourceServer`, and then add the dependency?

Reply:
It would be -

```ts
...
scopes: [
  OAuthScope.resourceServer(server, '*'),
],
```

I believe you can get the `users/` from the resource server's id.

> We have to check scopes, see which are of type resourceServer, and then add the dependency?

I don't follow what your questions is. Nevertheless, the implementation will be like this -

```ts
export class OAuthScope {
  public static resourceServer(server: UserPoolResourceServer, scopeName: string) {
    return new OAuthScope(`${server.userPoolResourceServerId}/${scopeName}`);
  }
}
```

Reply:
I see, i wasn't sure if just passing it into the function implicitly creates a dependency at the cloudformation level. Thanks for confirming